File: skincare-routines/chat-api/chat_api/tools/registry.py
"""Registry of site tools. Declarations are rebuilt per dataset version so enums track the live manifest."""
from __future__ import annotations

import json
import time

from .base import Tool, ToolContext, ToolError
from .catalog import GetReferenceCeiling, GetRoutines, GetScoringMethod, GetSiteOverview, ListCategories
from .knowledge import GetIngredientKnowledge
from .products import CompareProducts, GetCategoryFilters, GetProduct, GetTopProducts, SearchProducts

ALL_TOOLS: list[Tool] = [
    GetSiteOverview(), ListCategories(), GetScoringMethod(), GetReferenceCeiling(), GetRoutines(),
    SearchProducts(), GetTopProducts(), GetCategoryFilters(), GetProduct(), CompareProducts(), GetIngredientKnowledge(),
]


class ToolRegistry:
    def __init__(self, tools: list[Tool] | None = None):
        self._tools = {t.name: t for t in (tools or ALL_TOOLS)}
        self._declared_for: str | None = None
        self._declarations: list[dict] = []

    def names(self) -> list[str]:
        return list(self._tools)

    def declarations(self, manifest: dict) -> list[dict]:
        version = str(manifest.get("generatedAt"))
        if version != self._declared_for:
            self._declarations = [t.declaration(manifest) for t in self._tools.values()]
            self._declared_for = version
        return self._declarations

    async def execute(self, name: str, args: dict, ctx: ToolContext) -> tuple[dict, float]:
        """Run a tool; tool-level problems come back as an `error` payload the model can read, never as a crash."""
        started = time.perf_counter()
        tool = self._tools.get(name)
        if not tool:
            return {"error": f"Unknown tool {name}"}, 0.0
        try:
            result = await tool.run(args, ctx)
        except ToolError as exc:
            result = {"error": str(exc)}
        return result, (time.perf_counter() - started) * 1000


def size_of(payload: dict) -> int:
    return len(json.dumps(payload, ensure_ascii=False))

File: skincare-routines/src/chat/local/tools/registry.ts
/** Registry of site tools. Declarations are rebuilt per dataset version so enums track the live manifest. */
import type { Manifest } from '../../../lib/types';
import { declarationOf, ToolError, type Json, type Tool, type ToolContext, type ToolDeclaration } from './base';
import { getReferenceCeiling, getRoutines, getScoringMethod, getSiteOverview, listCategories } from './catalog';
import { getIngredientKnowledge } from './knowledge';
import { compareProducts, getCategoryFilters, getProduct, getTopProducts, searchProducts } from './products';
import { proposeRoutineSteps } from './routine';

export const ALL_TOOLS: Tool[] = [
  getSiteOverview, listCategories, getScoringMethod, getReferenceCeiling, getRoutines,
  searchProducts, getTopProducts, getCategoryFilters, getProduct, compareProducts, getIngredientKnowledge, proposeRoutineSteps,
];

export class ToolRegistry {
  private readonly tools: Map<string, Tool>;
  private declaredFor: string | null = null;
  private declared: ToolDeclaration[] = [];

  constructor(tools: Tool[] = ALL_TOOLS) { this.tools = new Map(tools.map((t) => [t.name, t])); }

  names() { return [...this.tools.keys()]; }

  surfaces(name: string) { return this.tools.get(name)?.surface === true; }

  declarations(manifest: Manifest): ToolDeclaration[] {
    if (manifest.generatedAt !== this.declaredFor) {
      this.declared = [...this.tools.values()].map((t) => declarationOf(t, manifest));
      this.declaredFor = manifest.generatedAt;
    }
    return this.declared;
  }

  /** Run a tool; tool-level problems come back as an `error` payload the model can read, never as a crash. */
  async execute(name: string, args: Json, ctx: ToolContext): Promise<{ result: Json; ms: number }> {
    const started = performance.now();
    const tool = this.tools.get(name);
    if (!tool) return { result: { error: `Unknown tool ${name}` }, ms: 0 };
    let result: Json;
    try {
      result = await tool.run(args, ctx);
    } catch (err) {
      if (!(err instanceof ToolError)) throw err;
      result = { error: err.message };
    }
    return { result, ms: performance.now() - started };
  }
}

export const sizeOf = (payload: Json) => JSON.stringify(payload).length;
